code/python/flw.py

```diff
@@ -0,0 +1,204 @@
+                                         config['b'])
+            new_follower.leader = id
+            leaders[id]['followers'].append(new_follower)
+            pop.append(new_follower)
+    # create walker agents
+    for _ in range(n_followers):
+        new_walker = init_location(config['dimension'], config['a'],
+                                   config['b'])
+        new_walker.leader = None
+        pop.append(new_walker)
+
+    return pop, leaders
+
+
+def update(agent, leaders, phi1=2, phi2=1, phi3=20):
+
+    if agent.leader is not None and agent == leaders[agent.leader]['agent']:  # leader?
+    #    u = (random.uniform(0, phi2) for _ in range(len(agent.position)))
+    #   new_position = list(map(operator.add, agent.position, u))
+    #   agent.position[:] = new_position
+        pass
+    elif agent.leader is not None:  # follower?
+        e = list(random.uniform(0, phi1) for _ in range(len(agent.position)))
+        v = list(random.uniform(0, phi2) for _ in range(len(agent.position)))
+        # e (xl - xi)
+        v_e = list(map(operator.mul, e, map(operator.sub,
+                                       leaders[agent.leader]['agent'].position,
+                                       agent.position)))
+        new_position = list(map(operator.add, agent.position,
+                                v_e))
+        # xi + e(xl - xi) + v
+        # new_position = list(map(operator.add, agent.position,
+        #                   map(operator.add, v_e, v)))
+
+        agent.position[:] = new_position
+
+    elif agent.leader is None:  # walker
+        w = list(random.uniform(0, random.choice(step_sigmas)) for _ in range(len(agent.position)))
+        new_position = list(map(operator.add, agent.position, w))
+        agent.position[:] = new_position
+
+    for i, position in enumerate(agent.position):
+        if abs(position) < agent.min:
+            agent.position[i] = math.copysign(agent.min, position)
+        elif abs(position) > agent.max:
+            agent.position[i] = math.copysign(agent.max, position)
+
+def print_pool(g, pool, leaders):
+    print("Gen :", g)
+    for agent_id in leaders:
+        print("leader:", agent_id, leaders[agent_id]['agent'])
+        for f in leaders[agent_id]['followers']:
+            print(f)
+    print("Walkers:")
+    for walker in [agent for agent in pool if agent.leader is None]:
+        print(walker)
+            
+
+    print("#####")
+
+
+evaluate = rastrigin  # This can be an import or a dictionary like in NetLogo
+
+
+def main(config):
+    stats = tools.Statistics(lambda ind: ind.value)
+    stats.register("avg", np.mean)
+    stats.register("std", np.std)
+    stats.register("min", np.min)
+    stats.register("max", np.max)
+
+    logbook = tools.Logbook()
+    logbook.header = ["steps", "evals"] + stats.fields
+
+    pool, leaders = create_agents(config)
+    best_leader = None
+    best_solution = None
+
+    for g in range(config['n_gens']):
```

Comment:
We should try and modularize this a bit further. There's too much business logic, here, I'm afraid...

Reply:
Yes!  I was planing to do that, but uploading it to see if someone can help.

Reply:
I'll have a look